### PR TITLE
chunk: support encode/decode a Chunk

### PR DESCRIPTION
**Pull request by zz-jason**
## What have you changed? (mandatory)

add `Encode` and `Decode` function to encode/decode a `Chunk`

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

- unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Does this PR need to be added to the release notes? (mandatory)

No

## Benchmark result if necessary (optional)

Decode 1024 rows, each row has 4 integer columns, all integer values are `123`.

Here is the benchmarks of [`ReadRowsData()`](https://github.com/pingcap/tidb/blob/master/distsql/distsql.go#L175), which decodes the input byte slice row by row into a `Chunk`:
```sh
➜ go test -bench BenchmarkReadRowsData -run xx -count 5 --benchmem
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/distsql
BenchmarkReadRowsData-8            10000            198137 ns/op          172346 B/op          0 allocs/op
BenchmarkReadRowsData-8            10000            153931 ns/op          172346 B/op          0 allocs/op
BenchmarkReadRowsData-8            10000            153893 ns/op          172346 B/op          0 allocs/op
BenchmarkReadRowsData-8            10000            154092 ns/op          172346 B/op          0 allocs/op
BenchmarkReadRowsData-8            10000            157055 ns/op          172346 B/op          0 allocs/op
PASS
ok      github.com/pingcap/tidb/distsql 8.397s
```

Here is the benchmark of `DecodeToChunk`:
```sh
➜ go test -bench BenchmarkDecodeToChunk -run xx -count 5 --benchmem
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/distsql
BenchmarkDecodeToChunk-8         2000000               813 ns/op               0 B/op          0 allocs/op
BenchmarkDecodeToChunk-8         2000000               822 ns/op               0 B/op          0 allocs/op
BenchmarkDecodeToChunk-8         2000000               841 ns/op               0 B/op          0 allocs/op
BenchmarkDecodeToChunk-8         2000000               833 ns/op               0 B/op          0 allocs/op
BenchmarkDecodeToChunk-8         2000000               827 ns/op               0 B/op          0 allocs/op
PASS
ok      github.com/pingcap/tidb/distsql 12.584s
```

### CODE REVIEW

File: util/chunk/codec.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	buffer = append(buffer, lenBuffer[:4]...)
+
+	// encode nullCount
+	binary.BigEndian.PutUint32(lenBuffer, uint32(col.nullCount))
+	buffer = append(buffer, lenBuffer[:4]...)
+
+	// encode nullBitmap
+	binary.BigEndian.PutUint32(lenBuffer, uint32(len(col.nullBitmap)))
+	buffer = append(buffer, lenBuffer[:4]...)
+	buffer = append(buffer, col.nullBitmap...)
```
Comment (coocood) [inline util/chunk/codec.go]: We need padding to 64 bytes, according apache arrow spec 
https://arrow.apache.org/docs/memory_layout.html

File: util/chunk/codec.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+// limitations under the License.
+
+package chunk
+
+import (
+	"encoding/binary"
+)
+
+func Encode(chk *Chunk) []byte {
+	buffer := make([]byte, 0, chk.MemoryUsage())
```
Comment (coocood) [inline util/chunk/codec.go]: Need to add padding budget.

File: util/chunk/codec.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		} else {
+			buffer = encodeVarLenColumn(buffer, col)
+		}
+	}
+	return buffer
+}
+
+func encodeColumn(buffer, lenBuffer []byte, col *column) []byte {
+	// encode length
+	binary.BigEndian.PutUint32(lenBuffer, uint32(col.length))
```
Comment (coocood) [inline util/chunk/codec.go]: arrow use LittleEndian by default.

File: util/chunk/codec.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+func encodeVarLenColumn(buffer []byte, col *column) []byte {
+	colOffset := uint64(len(buffer))
+
+	// encode column
+	var lenBuffer [8]byte
+	buffer = append(buffer, lenBuffer[:]...)
+	buffer = encodeColumn(buffer, lenBuffer[:], col)
+
+	// encode offsets
```
Comment (coocood) [inline util/chunk/codec.go]: Offset should be encoded before value.

File: util/chunk/codec.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	numOffsetBytes := binary.BigEndian.Uint32(buffer[numEatenBytes:])
+	numEatenBytes += 4
+	col.offsets = make([]int32, 0, numOffsetBytes/4)
+	for numEatenBytes < numTotalBytes {
+		offset := binary.BigEndian.Uint32(buffer[numEatenBytes:])
+		col.offsets = append(col.offsets, int32(offset))
+		numEatenBytes += 4
+	}
+
+	return buffer[:numTotalBytes]
```
Comment (AndreMouche) [inline util/chunk/codec.go]: seems should return buffer[numTotalBytes:]?

File: util/chunk/codec.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+	// encode column
+	var lenBuffer [8]byte
+	buffer = append(buffer, lenBuffer[:]...)
+	buffer = encodeColumn(buffer, lenBuffer[:], col)
+
+	// encode offsets
+	binary.BigEndian.PutUint32(lenBuffer[:], uint32(4*len(col.offsets)))
+	buffer = append(buffer, lenBuffer[:4]...)
+	for i := range col.offsets {
```
Comment (coocood) [inline util/chunk/codec.go]: We can convert int32 slice to byte slice directly.

```go
func i32SliceToBytes(i32s []int32) []byte {
	if len(i32s) == 0 {
		return nil
	}
	var b []byte
	hdr := (*reflect.SliceHeader)(unsafe.Pointer(&b))
	hdr.Len = len(i32s) * 4
	hdr.Cap = hdr.Len
	hdr.Data = uintptr(unsafe.Pointer(&i32s[0]))
	return b
}
```

File: util/chunk/codec.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+		binary.BigEndian.PutUint32(lenBuffer[:], uint32(col.offsets[i]))
+		buffer = append(buffer, lenBuffer[:4]...)
+	}
+
+	// supplement the length info of this column in the front
+	numBytes := uint64(len(buffer)) - colOffset - 8
```
Comment (AndreMouche) [inline util/chunk/codec.go]: could we encode a flag with 1 byte to indicate whether it is a fixed type or not(may support more complex type in the future)? so the total length could be removed?

File: util/chunk/codec.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	chk := &Chunk{}
+	for len(buffer) > 0 {
+		col := &column{}
+		buffer = decodeColumn(buffer, col)
+		chk.columns = append(chk.columns, col)
+	}
+	return chk
+}
+
+func decodeColumn(buffer []byte, col *column) []byte {
```
Comment (coocood) [inline util/chunk/codec.go]: We need to know if this column is fixed length or variable length before decode.

File: util/chunk/codec.go (hunk 7 above)
Comment (coocood) [reply util/chunk/codec.go]: This information is meta, it's not part of arrow format, we have the schema of the chunk before decode it.

File: util/chunk/codec.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	colOffset := uint64(len(buffer))
+
+	// encode column
+	var lenBuffer [8]byte
+	buffer = append(buffer, lenBuffer[:]...)
+	buffer = encodeColumn(buffer, lenBuffer[:], col)
+
+	// supplement the length info of this column in the front
+	numBytes := uint64(len(buffer)) - colOffset - 8
+	binary.BigEndian.PutUint64(lenBuffer[:], numBytes)
```
Comment (lysu) [inline util/chunk/codec.go]: ditto?

File: util/chunk/codec.go (hunk 7 above)
Comment (AndreMouche) [reply util/chunk/codec.go]: So the total length could be removed?

**Comment by zz-jason** [comment]
@coocood PTAL the `encode` method

File: util/chunk/codec.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+func (c *Codec) calcNumPaddingBytes(numAppendedBytes int) int {
+	numTailBytes := numAppendedBytes & 0x7
+	return (8 - numTailBytes) & 0x7
+}
+
+func (c *Codec) Decode(buffer []byte) *Chunk {
+	chk := &Chunk{}
+	for len(buffer) > 0 {
+		col := &column{}
+		buffer = c.decodeColumn(buffer, col)
```
Comment (coocood) [inline util/chunk/codec.go]: Need to decode varlen column.

**Comment by zz-jason** [comment]
@lysu No need to, we use fixed length encoding method to speed up the performance, we don't have much these length values thus var-length encoding method can not save much network bandwidth.

**Comment by zz-jason** [comment]
@coocood PTAL

**Comment by lysu** [comment]
should we need a `version` flag in encode result?

**Comment by zz-jason** [comment]
@lysu maybe we can add it to the protobuf message?

**Comment by lysu** [comment]
okey~

File: util/chunk/codec.go [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+	"reflect"
+	"unsafe"
+
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/types"
+)
+
+// Codec is used to:
+// 1. encode a Chunk to a byte slice.
+// 2. decode a CHunk from a byte slice.
```
Comment (crazycs520) [inline util/chunk/codec.go]: s/CHunk/chunk/

File: util/chunk/codec.go [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+	copy(buffer[startOffset:], lenBuffer[:4])
+
+	return buffer
+}
+
+func (c *Codec) encodeVarLenColumn(buffer []byte, col *column) []byte {
+	startOffset := len(buffer)
+
+	// encode column
+	var lenBuffer [4]byte
```
Comment (coocood) [inline util/chunk/codec.go]: We don't need to encode this.
The last offset is the data length.

File: util/chunk/codec.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+	binary.LittleEndian.PutUint32(lenBuffer, uint32(col.nullCount))
+	buffer = append(buffer, lenBuffer[:4]...)
+
+	// encode nullBitmap
+	if col.nullCount > 0 {
+		binary.LittleEndian.PutUint32(lenBuffer, uint32(len(col.nullBitmap)))
+		buffer = append(buffer, lenBuffer[:4]...)
+		buffer = append(buffer, col.nullBitmap...)
+	}
+
```
Comment (coocood) [inline util/chunk/codec.go]: Can we encode offset here?

**Comment by zz-jason** [comment]
@coocood PTAL

File: util/chunk/codec.go [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+type Codec struct {
+	// colTypes is used to check whether a column is fixed sized and what the
+	// fixed size for every element.
+	// NOTE: It's only used for decoding.
+	colTypes []*types.FieldType
+
+	// notNullBitmap is used to populate the nullBitmap for every decoded
+	// column, which has zero null values.
+	// NOTE: It's only used for decoding.
+	notNullBitmap []byte
```
Comment (coocood) [inline util/chunk/codec.go]: We can use a 128 byte len global variable instead.

**Comment by zz-jason** [comment]
@coocood PTAL

File: util/chunk/codec.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+	}
+	return chk, buffer
+}
+
+// DecodeToChunk decodes a Chunk from a byte slice, return the remained unused bytes.
+func (c *Codec) DecodeToChunk(buffer []byte, chk *Chunk) (remained []byte) {
+	for i := 0; i < len(chk.columns); i++ {
+		buffer = c.decodeColumn(buffer, chk.columns[i], i)
+	}
+	return buffer
```
Comment (lysu) [inline util/chunk/codec.go]: How to use returned `the remained unused bytes`? maybe take a test or example to show how to handle len(remained) > 0?

**Comment by coocood** [comment]
LGTM

**Comment by crazycs520** [comment]
LGTM

**Pull request merged**